elephant/gpfa.py

```diff
@@ -0,0 +1,407 @@
+                related to GP timescales by 'bin_width / sqrt(gamma)'
+            eps: ndarray of shape (1, #latent_vars)
+                GP noise variances
+            d: ndarray of shape (#units, 1)
+                observation mean
+            C: ndarray of shape (#units, #latent_vars)
+                mapping between the neuronal data space and the latent variable
+                space
+            R: ndarray of shape (#units, #latent_vars)
+                observation noise covariance
+
+    seqs_train: np.recarray
+        Second return value of extract_trajectory().
+        Data structure, whose n-th entry (corresponding to the n-th
+        experimental trial) has fields
+            * trialId: int
+                unique trial identifier
+            * T: int
+                number of timesteps
+            * y: ndarray of shape (#units, #bins)
+                neural data
+            * xsm: ndarray of shape (#latent_vars, #bins)
+                posterior mean of latent variables at each time bin
+            * Vsm: ndarray of shape (#latent_vars, #latent_vars, #bins)
+                posterior covariance between latent variables at each
+                timepoint
+            * VsmGP: ndarray of shape (#bins, #bins, #latent_vars)
+                posterior covariance over time for each latent variable
+
+    seqs_test: np.recarray, optional
+        Data structure of test dataset.
+        Default is None.
+
+    Returns
+    -------
+
+    params_est : dict
+        Estimated model parameters, including `Corth`, obtained by
+        orthonormalizing the columns of C.
+    seqs_train : np.recarray
+        Training data structure that contains the new field `xorth`,
+        the orthonormalized neural trajectories.
+    seqs_test : np.recarray
+        Test data structure that contains orthonormalized neural
+        trajectories in `xorth`, obtained using `params_est`.
+        When no test dataset is given, None is returned.
+
+
+    Raises
+    ------
+    ValueError
+        If `fit_info['kernSDList'] != kern_sd`.
+
+    """
+    C = params_est['C']
+    X = np.hstack(seqs_train['xsm'])
+    Xorth, Corth, _ = gpfa_util.orthogonalize(X, C)
+    seqs_train = gpfa_util.segment_by_trial(seqs_train, Xorth, 'xorth')
+
+    params_est['Corth'] = Corth
+
+    if seqs_test is not None:
+        print("Extracting neural trajectories for test data...\n")
+        seqs_test = gpfa_core.exact_inference_with_ll(seqs_test,
+                                                      params_est)
+        X = np.hstack(seqs_test['xsm'])
+        Xorth, Corth, _ = gpfa_util.orthogonalize(X, C)
+        seqs_test = gpfa_util.segment_by_trial(seqs_test, Xorth, 'xorth')
+
+    return params_est, seqs_train, seqs_test
+
+
+def gpfa(data, bin_size=20*pq.ms, x_dim=3, em_max_iters=500):
+    """
+    Prepares data and calls functions for extracting neural trajectories.
+
+    Parameters
+    ----------
+
+    data : list
```

Comment:
I also agree that the trial ID is useless here because it always goes as `range(len(data))`. I also added TODO below. Since this is GPFA-related issue only, we might change this behaviour now, not waiting for the next issue and corresponding fix (some people might start using it immediately). Discussion is needed.

Reply:
The trial ID is inherited from the original MATLAB code, and the idea behind this would be that trial IDs can start from arbitrary number instead of 0 or 1, and there could even be missing numbers in between.
I personally don't think this feature is necessary, and if we get rid of this, the input structure would be much simpler.